### ISSUE TITLE
Debian 10 cloud-init not properly intializing

### ISSUE THREAD

**Issue by berkutta** [giovtorres/kvm-install-vm#57]
I'm facing the issue that debian9 is working perfectly, but debian10 is not working. On debian10 the hostname is not set and also logging in via ssh is not working/enabled. I'm not sure if this is an issue with this script/debian butt image or my local setup.

I'm using Arch Linux with QEMU 4.2.0 / virt-install 2.2.1

![Screenshot from 2020-04-13 22-13-29](https://user-images.githubusercontent.com/11159719/79157372-c03d4100-7dd4-11ea-9383-d290db9d94a7.png)

**Comment by giovtorres** [comment]
I am also able to replicate. It appears this particular image is not getting instantiated the same way Debian 9 is. The SSH host keys aren't created, which explains the Connection Reset error. Additionally,  the `debian` user doesn't seem to exist on the image after boot.

@meffie Did you run into this as well, or had to make any changes to get Debian 10 to work?

**Comment by meffie** [comment]
I normally a set of pregenerated ssh keys, maybe that's why I did not see this issue.

$ cat .kivrc 
PUBKEY=/home/mmeffie/.ssh/mmeffie-sna2.pub
BRIDGE=vlbr0

$ kvm-install-vm create -t debian10 deb10
- Copying cloud image (debian-10-openstack-amd64.qcow2) ... OK
- Generating ISO for cloud-init ... OK
- Creating storage pool ... OK
- Installing the domain ... OK
- Cleaning up cloud-init files ... OK
- MAC address: 52:54:00:22:e8:1e
- Waiting for domain to get an IP address ... OK

- Checking for 192.168.123.35 in known_hosts file
- No entries found for 192.168.123.35

- SSH to deb10: 'ssh debian@192.168.123.35' or 'ssh debian@deb10'
- Console at spice://localhost:5912
- DONE
$ ssh debian@192.168.123.35 uname -a
The authenticity of host '192.168.123.35 (192.168.123.35)' can't be established.
ED25519 key fingerprint is SHA256:osD0R96TFO61S71rj4idqrEWCx64qWYyU0DCgaC71dk.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '192.168.123.35' (ED25519) to the list of known hosts.
Linux deb10 4.19.0-8-cloud-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) x86_64 GNU/Linux

**Comment by meffie** [comment]
I removed my PUBKEY setting, and created default keys, but it still works for me. My host kvm is ubuntu18.04, could that matter?

    $ cat ~/.kivrc 
    BRIDGE=vlbr0
    $ kvm-install-vm create -t debian10 deb10b
    ERR: Please generate an SSH keypair using 'ssh-keygen -t rsa' or specify one with the -k flag.

    $ ssh-keygen -t rsa
    Generating public/private rsa key pair.
    <snip>

    $ kvm-install-vm create -t debian10 deb10b
    - Copying cloud image (debian-10-openstack-amd64.qcow2) ... OK
    - Generating ISO for cloud-init ... OK
    - Creating storage pool ... OK
    - Installing the domain ... OK
    - Cleaning up cloud-init files ... OK
    - MAC address: 52:54:00:bd:4b:cc
    - Waiting for domain to get an IP address ... OK
    
    - Checking for 192.168.123.127 in known_hosts file
    - No entries found for 192.168.123.127
    
    - SSH to deb10b: 'ssh debian@192.168.123.127' or 'ssh debian@deb10b'
    - Console at spice://localhost:5913
    - DONE

    $ ssh debian@192.168.123.127
    The authenticity of host '192.168.123.127 (192.168.123.127)' can't be established.
    ED25519 key fingerprint is SHA256:M1I5qzrCMxB5OwT6Rlvm/RDFIdWCWdCGFmpncBaqrjI.
    Are you sure you want to continue connecting (yes/no)? yes 
    Warning: Permanently added '192.168.123.127' (ED25519) to the list of known hosts.
    Linux deb10b 4.19.0-8-cloud-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) x86_64
   
    <snip>
    $ cat /etc/debian_version 
    10.3

**Comment by meffie** [comment]
Asking a colleague to try to reproduce .

**Comment by berkutta** [comment]
Hmm, shouldn't there at least be a cloud-init message in the boot log?

**Comment by CheyenneWills** [comment]
Actually - the ssh port is open.  it's being reset..

**Comment by berkutta** [comment]
It shouldn't have anything to do with ssh. Cloud-init works with a virtual cdrom drive with the configuration in it.

**Comment by meffie** [comment]
Looks like my cached copy of the image was from March 28. Maybe the upstream image changed
since then?

    -rw-rw-r-- 1 mmeffie mmeffie 587613696 Mar 28 12:50 debian-10-openstack-amd64.qcow2

**Comment by meffie** [comment]
Downloaded the most recent image and it still works for me. Strange. cloud-init runs and everything seems fine.

**Comment by berkutta** [comment]
Seems also weird on Ubuntu 20.04. Will try Ubuntu 18.04 tomorrow.
```
ubuntu@focal0:~/kvm-install-vm$ ./kvm-install-vm create -t debian10 foo
[WARNING] foo already exists.  Do you want to overwrite foo [y/N]? y
- Destroying foo domain ... OK
- Undefining foo domain ... OK
- Deleting foo files ... OK
- Destroying foo storage pool ... OK
- Copying cloud image (debian-10-openstack-amd64.qcow2) ... OK
- Generating ISO for cloud-init ... OK
- Creating storage pool ... OK
- Installing the domain ... OK
- Cleaning up cloud-init files ... OK
- MAC address: 52:54:00:a6:ba:e8
- Waiting for domain to get an IP address ... OK

- Checking for 192.168.123.42 in known_hosts file
- No entries found for 192.168.123.42

- SSH to foo: 'ssh debian@192.168.123.42' or 'ssh debian@foo'
- Console at spice://localhost:5900
- DONE
ubuntu@focal0:~/kvm-install-vm$ ssh debian@192.168.123.42
Connection reset by 192.168.123.42 port 22
```

**Comment by CheyenneWills** [comment]
So this works under xen (with a gentoo dom0), but is failing on a gentoo kvm system.  On the kvm system debian 9 works.  Looking at the logs it appears that cloud-init is never kicking in. 

I've captured console logs (added a --console pty,log.file=... to the virt-install).  

I'll continue to look at this tomorrow

**Comment by berkutta** [comment]
Ok, I think I found the "issue". This hacky change "fixes" it https://github.com/berkutta/kvm-install-vm/commit/762d00dcb559bafbe746b354440200f3213924be

I don't know why this fixes it. But I can reproduce this issue on ubuntu20.04 hosts (ubuntu18.04 hots work fine). The issue is also reproducible on Arch Linux.

**Comment by meffie** [comment]
On Fri, 24 Apr 2020 01:52:42 -0700
Benjamin Marty <notifications@github.com> wrote:

> Ok, I think I found the "issue". This hacky change "fixes" it https://github.com/berkutta/kvm-install-vm/commit/762d00dcb559bafbe746b354440200f3213924be
> 
> I don't know why this fixes it. But I can reproduce this issue on ubuntu20.04 hosts (ubuntu18.04 hots work fine). The issue is also reproducible on Arch Linux.

Interesting! I can confirm my ubuntu 18.04 host still works with this change, for
at least debian10 guests, I've not tried others yet. 

I see this changes the virt-install --disk options for the cdrom from

    virt-install \
    ...
    --disk=deb10c-cidata.iso,device=cdrom  \ 

to:

   virt-install \
     ...
     --disk  \ 
     path=deb10b-cidata.iso  \

on my ubuntu 18.04 host.

-- 
Michael Meffie <mmeffie@sinenomine.net>

**Comment by CheyenneWills** [comment]
Confirm that the "hacky" change works for deb10 on gentoo/kvm.

Installing deb10 on a xen host continues to work as well.

**Comment by berkutta** [comment]
The root cause for all this trouble could be that newer libvirt versions use SATA for ejectable media instead of IDE. (Saw this difference when comparing the VM config between what virt-install generatres on ubuntu18.04 and ubuntu20.04).

**Comment by cy4n1c** [comment]
> Ok, I think I found the "issue". This hacky change "fixes" it [berkutta@762d00d](https://github.com/berkutta/kvm-install-vm/commit/762d00dcb559bafbe746b354440200f3213924be)

Confirming that this commit fixes the issue on Debian 10 host.

**Comment by giovtorres** [comment]
Thanks everyone for testing this fix! @berkutta would you like to submit a PR?

**Comment by berkutta** [comment]
sure, PR is ready, did clean up the code.

**Comment by giovtorres** [comment]
Closed by #59
